File: examples/ts/bsv20-token/BSV20Token.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract, ALICE, signTestMessage } from 'runar-testing';
import { compile } from 'runar-compiler';
import type { RunarArtifact } from 'runar-ir-schema';
import { RunarContract } from '../../../packages/runar-sdk/src/contract.js';
import {
  parseInscriptionEnvelope,
} from '../../../packages/runar-sdk/src/ordinals/envelope.js';
import { BSV20 } from '../../../packages/runar-sdk/src/ordinals/bsv20.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'BSV20Token.runar.ts'), 'utf8');

const SIG = signTestMessage(ALICE.privKey);

/** Convert hex to UTF-8 string. */
function hexToUtf8(hex: string): string {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.slice(i, i + 2), 16);
  }
  return new TextDecoder().decode(bytes);
}

function compileContract(): RunarArtifact {
  const result = compile(source, { fileName: 'BSV20Token.runar.ts' });
  if (!result.success || !result.artifact) {
    const errors = result.diagnostics
      .filter((d) => d.severity === 'error')
      .map((d) => d.message)
      .join('\n  ');
    throw new Error(`BSV20Token compile failed:\n  ${errors}`);
  }
  return result.artifact;
}

describe('BSV20Token', () => {
  describe('business logic', () => {
    it('accepts a valid unlock', () => {
      const contract = TestContract.fromSource(source, { pubKeyHash: ALICE.pubKeyHash });
      const result = contract.call('unlock', { sig: SIG, pubKey: ALICE.pubKey });
      expect(result.success).toBe(true);
    });
  });

  describe('compilation', () => {
    it('compiles to a valid artifact', () => {
      const artifact = compileContract();
      expect(artifact.contractName).toBe('BSV20Token');
    });
  });

  describe('BSV-20 deploy inscription', () => {
    it('creates a deploy inscription with correct JSON', () => {
      const artifact = compileContract();
      const inscription = BSV20.deploy({ tick: 'RUNAR', max: '21000000', lim: '1000' });
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);
      contract.withInscription(inscription);

      const lockingScript = contract.getLockingScript();
      const parsed = parseInscriptionEnvelope(lockingScript);

      expect(parsed).not.toBeNull();
      expect(parsed!.contentType).toBe('application/bsv-20');

      const json = JSON.parse(hexToUtf8(parsed!.data));
      expect(json.p).toBe('bsv-20');
      expect(json.op).toBe('deploy');
      expect(json.tick).toBe('RUNAR');
      expect(json.max).toBe('21000000');
      expect(json.lim).toBe('1000');
    });

    it('creates a deploy inscription with decimals', () => {
      const artifact = compileContract();
      const inscription = BSV20.deploy({ tick: 'USDT', max: '100000000', dec: '8' });
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);
      contract.withInscription(inscription);

      const lockingScript = contract.getLockingScript();
      const parsed = parseInscriptionEnvelope(lockingScript);

      const json = JSON.parse(hexToUtf8(parsed!.data));
      expect(json.dec).toBe('8');
    });
  });

  describe('BSV-20 mint inscription', () => {
    it('creates a mint inscription with correct JSON', () => {
      const artifact = compileContract();
      const inscription = BSV20.mint({ tick: 'RUNAR', amt: '1000' });
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);
      contract.withInscription(inscription);

      const lockingScript = contract.getLockingScript();
      const parsed = parseInscriptionEnvelope(lockingScript);

      expect(parsed).not.toBeNull();
      expect(parsed!.contentType).toBe('application/bsv-20');

      const json = JSON.parse(hexToUtf8(parsed!.data));
      expect(json.p).toBe('bsv-20');
      expect(json.op).toBe('mint');
      expect(json.tick).toBe('RUNAR');
      expect(json.amt).toBe('1000');
    });
  });

  describe('BSV-20 transfer inscription', () => {
    it('creates a transfer inscription with correct JSON', () => {
      const artifact = compileContract();
      const inscription = BSV20.transfer({ tick: 'RUNAR', amt: '50' });
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);
      contract.withInscription(inscription);

      const lockingScript = contract.getLockingScript();
      const parsed = parseInscriptionEnvelope(lockingScript);

      expect(parsed).not.toBeNull();
      expect(parsed!.contentType).toBe('application/bsv-20');

      const json = JSON.parse(hexToUtf8(parsed!.data));
      expect(json.p).toBe('bsv-20');
      expect(json.op).toBe('transfer');
      expect(json.tick).toBe('RUNAR');
      expect(json.amt).toBe('50');
    });
  });

  describe('round-trip', () => {
    it('inscription survives fromUtxo round-trip', () => {
      const artifact = compileContract();
      const inscription = BSV20.deploy({ tick: 'TEST', max: '1000' });
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);
      contract.withInscription(inscription);

      const lockingScript = contract.getLockingScript();

      const reconnected = RunarContract.fromUtxo(artifact, {
        txid: '00'.repeat(32),
        outputIndex: 0,
        satoshis: 1,
        script: lockingScript,
      });

      expect(reconnected.inscription).not.toBeNull();
      expect(reconnected.inscription!.contentType).toBe('application/bsv-20');

      const json = JSON.parse(hexToUtf8(reconnected.inscription!.data));
      expect(json.p).toBe('bsv-20');
      expect(json.op).toBe('deploy');
      expect(json.tick).toBe('TEST');
      expect(json.max).toBe('1000');
    });
  });
});

File: examples/go/bsv20-token/BSV20Token_test.go
package contract

import (
	"encoding/hex"
	"encoding/json"
	"testing"

	"github.com/icellan/runar/compilers/go/compiler"
	runar "github.com/icellan/runar/packages/runar-go"
)

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

// compileContract compiles BSV20Token.runar.go end-to-end and returns a
// runar.RunarArtifact suitable for driving the SDK. The compiler package's
// Artifact is JSON-equivalent to runar.RunarArtifact, so we round-trip via
// JSON to cross the package boundary.
func compileContract(t *testing.T) *runar.RunarArtifact {
	t.Helper()
	a, err := compiler.CompileFromSource("BSV20Token.runar.go")
	if err != nil {
		t.Fatalf("BSV20Token compile failed: %v", err)
	}
	raw, err := compiler.ArtifactToJSON(a)
	if err != nil {
		t.Fatalf("artifact marshal failed: %v", err)
	}
	var out runar.RunarArtifact
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatalf("artifact unmarshal failed: %v", err)
	}
	return &out
}

// hexToBytes decodes a hex string to bytes, failing the test on error.
func hexToBytes(t *testing.T, h string) []byte {
	t.Helper()
	b, err := hex.DecodeString(h)
	if err != nil {
		t.Fatalf("bad hex %q: %v", h, err)
	}
	return b
}

// parseInscriptionJSON decodes the UTF-8 JSON payload from an inscription's
// hex-encoded data field into a map.
func parseInscriptionJSON(t *testing.T, dataHex string) map[string]interface{} {
	t.Helper()
	var m map[string]interface{}
	if err := json.Unmarshal(hexToBytes(t, dataHex), &m); err != nil {
		t.Fatalf("parse inscription JSON: %v", err)
	}
	return m
}

// ---------------------------------------------------------------------------
// Business logic tests
// ---------------------------------------------------------------------------

func TestBSV20Token_Unlock(t *testing.T) {
	pk := runar.Alice.PubKey
	c := &BSV20Token{PubKeyHash: runar.Hash160(pk)}
	c.Unlock(runar.SignTestMessage(runar.Alice.PrivKey), pk)
}

func TestBSV20Token_Unlock_WrongKey(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure for wrong public key")
		}
	}()
	pk := runar.Alice.PubKey
	c := &BSV20Token{PubKeyHash: runar.Hash160(pk)}
	c.Unlock(runar.SignTestMessage(runar.Bob.PrivKey), runar.Bob.PubKey)
}

// ---------------------------------------------------------------------------
// Rúnar compile check
// ---------------------------------------------------------------------------

func TestBSV20Token_Compile(t *testing.T) {
	if err := runar.CompileCheck("BSV20Token.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

func TestBSV20Token_CompilesToValidArtifact(t *testing.T) {
	a := compileContract(t)
	if a.ContractName != "BSV20Token" {
		t.Errorf("expected contractName BSV20Token, got %q", a.ContractName)
	}
}

// ---------------------------------------------------------------------------
// BSV-20 deploy inscription
// ---------------------------------------------------------------------------

func TestBSV20Token_DeployInscription(t *testing.T) {
	a := compileContract(t)
	lim := "1000"
	inscription := runar.BSV20Deploy("RUNAR", "21000000", &lim, nil)
	contract := runar.NewRunarContract(a, []interface{}{runar.Alice.PubKeyHash})
	contract.WithInscription(inscription)

	lockingScript := contract.GetLockingScript()
	parsed := runar.ParseInscriptionEnvelope(lockingScript)
	if parsed == nil {
		t.Fatal("expected inscription envelope, got nil")
	}
	if parsed.ContentType != "application/bsv-20" {
		t.Errorf("contentType = %q, want application/bsv-20", parsed.ContentType)
	}
	m := parseInscriptionJSON(t, parsed.Data)
	if m["p"] != "bsv-20" || m["op"] != "deploy" || m["tick"] != "RUNAR" || m["max"] != "21000000" || m["lim"] != "1000" {
		t.Errorf("unexpected deploy payload: %+v", m)
	}
}

func TestBSV20Token_DeployInscription_WithDecimals(t *testing.T) {
	a := compileContract(t)
	dec := "8"
	inscription := runar.BSV20Deploy("USDT", "100000000", nil, &dec)
	contract := runar.NewRunarContract(a, []interface{}{runar.Alice.PubKeyHash})
	contract.WithInscription(inscription)

	lockingScript := contract.GetLockingScript()
	parsed := runar.ParseInscriptionEnvelope(lockingScript)
	if parsed == nil {
		t.Fatal("expected inscription envelope, got nil")
	}
	m := parseInscriptionJSON(t, parsed.Data)
	if m["dec"] != "8" {
		t.Errorf("dec = %v, want 8", m["dec"])
	}
}

// ---------------------------------------------------------------------------
// BSV-20 mint inscription
// ---------------------------------------------------------------------------

func TestBSV20Token_MintInscription(t *testing.T) {
	a := compileContract(t)
	inscription := runar.BSV20Mint("RUNAR", "1000")
	contract := runar.NewRunarContract(a, []interface{}{runar.Alice.PubKeyHash})
	contract.WithInscription(inscription)

	lockingScript := contract.GetLockingScript()
	parsed := runar.ParseInscriptionEnvelope(lockingScript)
	if parsed == nil {
		t.Fatal("expected inscription envelope, got nil")
	}
	if parsed.ContentType != "application/bsv-20" {
		t.Errorf("contentType = %q, want application/bsv-20", parsed.ContentType)
	}
	m := parseInscriptionJSON(t, parsed.Data)
	if m["p"] != "bsv-20" || m["op"] != "mint" || m["tick"] != "RUNAR" || m["amt"] != "1000" {
		t.Errorf("unexpected mint payload: %+v", m)
	}
}

// ---------------------------------------------------------------------------
// BSV-20 transfer inscription
// ---------------------------------------------------------------------------

func TestBSV20Token_TransferInscription(t *testing.T) {
	a := compileContract(t)
	inscription := runar.BSV20Transfer("RUNAR", "50")
	contract := runar.NewRunarContract(a, []interface{}{runar.Alice.PubKeyHash})
	contract.WithInscription(inscription)

	lockingScript := contract.GetLockingScript()
	parsed := runar.ParseInscriptionEnvelope(lockingScript)
	if parsed == nil {
		t.Fatal("expected inscription envelope, got nil")
	}
	if parsed.ContentType != "application/bsv-20" {
		t.Errorf("contentType = %q, want application/bsv-20", parsed.ContentType)
	}
	m := parseInscriptionJSON(t, parsed.Data)
	if m["p"] != "bsv-20" || m["op"] != "transfer" || m["tick"] != "RUNAR" || m["amt"] != "50" {
		t.Errorf("unexpected transfer payload: %+v", m)
	}
}

// ---------------------------------------------------------------------------
// Round-trip: inscription survives FromUtxo
// ---------------------------------------------------------------------------

func TestBSV20Token_InscriptionFromUtxoRoundTrip(t *testing.T) {
	a := compileContract(t)
	inscription := runar.BSV20Deploy("TEST", "1000", nil, nil)
	contract := runar.NewRunarContract(a, []interface{}{runar.Alice.PubKeyHash})
	contract.WithInscription(inscription)

	lockingScript := contract.GetLockingScript()
	reconnected := runar.FromUtxo(a, runar.UTXO{
		Txid:        "0000000000000000000000000000000000000000000000000000000000000000",
		OutputIndex: 0,
		Satoshis:    1,
		Script:      lockingScript,
	})
	if reconnected.GetInscription() == nil {
		t.Fatal("expected inscription after FromUtxo")
	}
	if reconnected.GetInscription().ContentType != "application/bsv-20" {
		t.Errorf("contentType = %q, want application/bsv-20", reconnected.GetInscription().ContentType)
	}
	m := parseInscriptionJSON(t, reconnected.GetInscription().Data)
	if m["p"] != "bsv-20" || m["op"] != "deploy" || m["tick"] != "TEST" || m["max"] != "1000" {
		t.Errorf("unexpected round-trip payload: %+v", m)
	}
}
